contracts/reserve/ExchangeHelper.sol

```diff
@@ -0,0 +1,313 @@
+			);
+			require(
+				_buyWith.transferFrom(
+					msg.sender,
+					address(this),
+					_tokenAmount
+				) == true,
+				"transferFrom failed, make sure you approved input token transfer"
+			);
+		}
+
+		uint256 result;
+		if (address(_buyWith) == cDaiAddress) {
+			require(
+				_buyWith.transferFrom(
+					msg.sender,
+					address(reserve),
+					_tokenAmount
+				) == true,
+				"transferFrom failed, make sure you approved input token transfer"
+			);
+			result = reserve.buy(_tokenAmount, _minReturn, receiver);
+		} else if (address(_buyWith) == daiAddress) {
+			result = _cdaiMintAndBuy(_tokenAmount, _minReturn, receiver);
+		} else {
+			address[] memory path = new address[](2);
+
+			path[0] = address(_buyWith);
+			path[1] = daiAddress;
+			uint256[] memory swap;
+			if (withETH) {
+				swap = uniswapContract.swapExactETHForTokens{
+					value: msg.value
+				}(_minDAIAmount, path, address(this), block.timestamp);
+			} else {
+				_buyWith.approve(address(uniswapContract), _tokenAmount);
+				swap = uniswapContract.swapExactTokensForTokens(
+					_tokenAmount,
+					_minDAIAmount,
+					path,
+					address(this),
+					block.timestamp
+				);
+			}
+			uint256 dai = swap[1];
+			require(dai > 0, "token selling failed");
+
+			result = _cdaiMintAndBuy(dai, _minReturn, receiver);
+		}
+
+		emit TokenPurchased(
+			msg.sender,
+			address(_buyWith),
+			_tokenAmount,
+			result
+		);
+
+		return result;
+	}
+
+	/**
+	 * @dev Converts GD tokens to `sellTo` tokens and update the bonding curve params.
+	 * `sell` occurs only if the token return is above the given minimum. Notice that
+	 * there is a contribution amount from the given GD that remains in the reserve relative to user amount of GDX credits.
+	 * MUST call G$ `approve` prior to this action to allow this
+	 * contract to accomplish the conversion.
+	 * @param _sellTo The tokens that will be received after the conversion if address equals 0x0 then sell to ETH
+	 * @param _gdAmount The amount of GD tokens that should be converted to `_sellTo` tokens
+	 * @param _minReturn The minimum allowed `sellTo` tokens return
+	 * @param _minTokenReturn The mininmum dai out amount from Exchange swap function
+	 * @param _targetAddress address of _sellTo token recipient if different than msg.sender
+	 * @return (tokenReturn) How much `sellTo` tokens were transferred
+	 */
+	function sell(
+		ERC20 _sellTo,
+		uint256 _gdAmount,
+		uint256 _minReturn,
+		uint256 _minTokenReturn,
+		address _targetAddress
+	) public nonReentrant returns (uint256) {
+		address receiver = _targetAddress == address(0x0)
+			? msg.sender
+			: _targetAddress;
+
+		uint256 result;
+		uint256 contributionAmount;
+		GoodReserveCDai reserve = GoodReserveCDai(
+			nameService.getAddress("RESERVE")
+		);
+		IGoodDollar(nameService.getAddress("GOODDOLLAR")).burnFrom(
+			msg.sender,
+			_gdAmount
+		);
+		(result, contributionAmount) = reserve.sell(
+			_gdAmount,
+			_minReturn,
+			address(this),
+			msg.sender
```

Comment:
if sellTo is cdai this should go directly to receiver

Reply:
Right , replaced it and cancelled below if for cdai